lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/FunctionalChaosValidation.cxx

```diff
@@ -0,0 +1,187 @@
+    const FunctionalChaosResult & functionalChaosResult,
+    const CrossValidationMethod cvMethod, 
+    const UnsignedInteger & kParameter)
+{
+  const Sample residualsSample(functionalChaosResult.getSampleResiduals());
+  const Sample inputSample(functionalChaosResult.getInputSample());
+  const UnsignedInteger sampleSize = inputSample.getSize();
+  const FunctionCollection reducedBasis(functionalChaosResult.getReducedBasis());
+  const UnsignedInteger reducedBasisSize = reducedBasis.getSize();
+  if (reducedBasisSize > sampleSize)
+    throw InvalidArgumentException(HERE) << "Error: the sample size is: " << sampleSize << 
+      " which is lower than the basis size: " << reducedBasisSize;
+  const Sample outputSample(functionalChaosResult.getOutputSample());
+  const UnsignedInteger outputDimension = outputSample.getDimension();
+  const Function metamodel(functionalChaosResult.getMetaModel());
+  const Function transformation(functionalChaosResult.getTransformation());
+  const Sample standardSample(transformation(inputSample));
+  DesignProxy designProxy(standardSample, reducedBasis);
+  Indices allIndices(reducedBasisSize);
+  allIndices.fill();
+  const MatrixImplementation designMatrix(designProxy.computeDesign(allIndices));
+  // The method name is set to the default one, given by ResourceMap
+  const String methodName(ResourceMap::GetAsString("LeastSquaresExpansion-DecompositionMethod"));
+  LeastSquaresMethod leastSquaresMethod(LeastSquaresMethod::Build(methodName, designProxy, allIndices));
+  leastSquaresMethod.update(Indices(0), allIndices, Indices(0));
+  Sample cvPredictions(sampleSize, outputDimension);
+  if (cvMethod == LEAVEONEOUT)
+  {
+    const Point diagonalH = leastSquaresMethod.getHDiag();
+    for (UnsignedInteger j = 0; j < outputDimension; ++j)
+      for (UnsignedInteger i = 0; i < sampleSize; ++i)
+        cvPredictions(i, j) = outputSample(i, j) - residualsSample(i, j) / (1.0 - diagonalH[i]);
+  }
+  else if (cvMethod == KFOLD)
+  {
+    const SymmetricMatrix projectionMatrix(leastSquaresMethod.getH());
+    // Compute K-Fold error
+    KFoldSplitter splitter(sampleSize, kParameter);
+    for (UnsignedInteger foldIndex = 0; foldIndex < kParameter; ++foldIndex)
+    {
+      Indices indicesTest;
+      const Indices indicesTrain(splitter.generate(indicesTest));
+      const UnsignedInteger foldSize = indicesTest.getSize();
+      SymmetricMatrix projectionKFoldMatrix(foldSize);
+      for (UnsignedInteger i1 = 0; i1 < foldSize; ++i1)
+        for (UnsignedInteger i2 = 0; i2 < 1 + i1; ++i2)
+          projectionKFoldMatrix(i1, i2) = projectionMatrix(indicesTest[i1], indicesTest[i2]);
+      const IdentityMatrix identityMatrix(foldSize);
+      const SymmetricMatrix reducedMatrix(identityMatrix - projectionKFoldMatrix);
+      const Sample residualsSampleKFoldTest(residualsSample.select(indicesTest));
+      Matrix multipleRightHandSide(foldSize, outputDimension);
```

Comment:
If foldSize does not change, this matrix should be moved outside of the loop.

Reply:
That is a correct suggestion in general, but does not apply here. In the case where the number of folds does not divide the sample size, then all folds do not have the same size. For example, if $n = 17$ and we use $k = 3$ folds. Then the folds sizes are 6, 6, 5. If that would become a performance issue, this code could be changed so that a slightly larger matrix is defined out of the loop, then the necessary rows are addressed within the loop. In the meantime, I suggest to keep the code as simple as possible. This is because I do not think that the analytical K-fold CV method will not have many uses, given the bias superiority of the LOO analytical CV.